Implement non-POD field access

darichey opened the pull request:
Hi, this is some work on implementing field access for non-POD types. The approach is based on the ideas from #53 (roughly, it seems a lot has changed since that was written). We get in early (immediately after parsing) in the analysis and insert synthetic C++ methods for accessing the fields.

It's pretty rough, but I wanted to get some early feedback on the general approach, and ask about a few issues. In particular, I'm a bit stumped by a problem presented by the `test_non_pod_with_non_copyable_member` test:
```cpp
struct Foo {
  Foo(const Foo &foo) = delete;
  int x;
};
struct Bar {
  Foo foo;
};
```
With these changes, we will attempt to insert an accessor like:
```cpp
Foo Bar_get_foo(const Bar &bar) {
  return bar.foo;
}
```
This will fail, because we cannot copy `foo`. What should we do here? This might be related to my next point.

I was also considering the fact that in C++ we have several options for accessing a field, for example:
```cpp
struct Baz {
  int x;
}
```
```cpp
Baz baz{};
int x1 = baz.x;
int* x2 = &baz.x;
int& x3 = baz.x;
```
Is there a nice way we could support all of these?

I would greatly appreciate any feedback. Thanks!

adetaylor commented:
Thanks a lot for raising this and figuring out where in the codebase to put everything. I am currently away for a while so I might not have a chance to do a thorough review straight away. But meanwhile I am excited you've done this. I have some ideas for the problems you've encountered but they are probably too nuanced for me to type on my phone, so expect a proper reply in a while. Thanks again!

adetaylor commented on engine/src/conversion/analysis/fun/mod.rs:
```diff
@@ -952,9 +952,19 @@ impl<'a> FnAnalyzer<'a> {
+                    (&rust_name[rust_name.find("get_").unwrap()..]).to_string()
+                } else {
+                    // Disambiguate overloads.
+                    predetermined_rust_name
```
I wonder if you should do this irrespective, in case `get_foo` to get a field `foo` conflicts with an existing method called `get_foo`

adetaylor replied:
(though I can see you're trying to avoid such conflicts elsewhere. Hmm. That's probably fine, then)

adetaylor commented on engine/src/conversion/analysis/struct_accessors.rs:
```diff
@@ -0,0 +1,142 @@
+use std::collections::HashSet;
```
Please add the same header/copyright comment as the other rs files

adetaylor commented on engine/src/conversion/analysis/struct_accessors.rs:
```diff
@@ -0,0 +1,142 @@
+use super::fun::function_wrapper::{CppFunctionBody, CppFunctionKind};
+
+pub(crate) fn add_field_accessors(
+    // TODO: this isn't used currently, but maybe there will be some option for doing this phase
```
I'd prefer to get rid of it, then. We can add it later if needs be.

adetaylor commented on engine/src/conversion/analysis/struct_accessors.rs:
```diff
@@ -0,0 +1,142 @@
+                        })
+                    }
+                }
+                // TODO: is it even possible for these to occur?
```
Probably not, but as a matter of policy I try to assume bindgen might give us weird input for reasons I haven't anticipated, so I just ignore such things rather than `todo!()`.

adetaylor commented on engine/src/conversion/analysis/struct_accessors.rs:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    // Don't generate accessors for cpp reference fields
+    // TODO: is this restriction necessary?
```
I think you're right to err on the side of paranoia for the first iteration here, and bail out in all circumstances where problems could conceivably occur. We can expand coverage later.

adetaylor commented on engine/src/conversion/analysis/struct_accessors.rs:
```diff
@@ -0,0 +1,142 @@
+    }
+
+    // Don't generate accessors for "fake" bindgen fields which wouldn't appear directly in the C++ struct
+    // TODO: is there a better way to tell which fields these are?
```
I don't know of a better way right now. In principle I like the idea of bindgen-specific knowledge being isolated into the `parse` bit of the codebase and I am gradually heading in that direction, but there's no consistency there yet. This is fine.

adetaylor commented on engine/src/conversion/analysis/struct_accessors.rs:
```diff
@@ -0,0 +1,142 @@
+}
+
+fn is_cpp_reference(attr: &syn::Attribute) -> bool {
+    return *attr == parse_quote! { #[cpp_semantics(reference)] }
```
Please could you add logic to `src/conversion/parse/bindgen_semantic_attributes.rs` to understand this attribute, then use that to do this attribute parsing? (It's fine with me, for now, if that code needs to be made visible outside `parse`)

darichey replied:
Just to confirm, did you mean you are ok with that method being public as it is here: https://github.com/google/autocxx/pull/1133/files#diff-887ed0eb29c1743994a938da92e9c5455536ef703017656b53b7f23417f77f18R79?

adetaylor commented on engine/src/conversion/api.rs:
```diff
@@ -221,6 +221,7 @@ pub(crate) enum SpecialMemberKind {
     MoveConstructor,
     Destructor,
     AssignmentOperator,
+    GeneratedAccessor,
```
Unfortunately I don't think you can add this here. I didn't add this bit of code, but AIUI "special member" has a specific meaning in C++ - https://cplusplus.com/doc/tutorial/classes2/ - and a generated accessor isn't one of them.

darichey replied:
Oops, of course. I was using this to identify which are the accessors that we generated during function analysis, specifically for giving nicer rust names. I'll need to come up with a different strategy. Do you think it would make sense to have the accessor generation add that info in the analysis, similar to how POD types are labelled?

adetaylor commented on engine/src/conversion/analysis/fun/mod.rs:
```diff
@@ -952,9 +952,19 @@ impl<'a> FnAnalyzer<'a> {
+                let rust_name = if matches!(
+                    fun.special_member,
+                    Some(SpecialMemberKind::GeneratedAccessor)
+                ) {
```
Could you explain why this `if` clause is necessary? What does the Rust name end up being otherwise?

darichey replied:
Without the `if`, the rust name will still have the struct name prefix (we can see this in the current test for now since I reverted this while we think of a better strategy for determining which accessors were generated: https://github.com/google/autocxx/pull/1133/files#diff-7d1a694c0fef109924a3e6d7c3462f13a4727cfa3b2a2f86620734ca3e1800fcR11511)

adetaylor reviewed (changes requested):
Great!

adetaylor commented:
General comments:

* There are some test failures that you should look into. #1134 is the only known test failure on main right now, so all the others are a consequence of this change.
* Thanks for doing the tedious CLA thing
* Regarding non-copyable things. Options:
  * (simpler) Initially provide accessors only for POD types. You should be able to look up which types are POD after we've called `analyze_pod_apis` which is pretty early in the analysis. Later than you're currently doing it, but not much later. (There are various existing bits of code which enumerate the POD types - see for example `build_pod_safe_type_set`)
  * (more complex) Add these accessors later once we've identified which types have `SpecialMemberKind::CopyConstructor` and friends. That's much later in the analysis, and would be a more ambitious change by far.
* Getting references/pointers to fields - my first instinct is not to do it as part of this PR but do it later (incremental changes FTW). But to avoid future naming conflicts, we should imagine what such future accessors might be called. I think we'd probably call them `get_foo_ref` or `get_ref_foo` (same with ptr) and thus it's OK if the current accessors are called `get_foo`

Thanks again for having a crack at this!

darichey commented:
Thanks a lot for the feedback. I believe I have addressed all of it!

> There are some test failures that you should look into

These were all related to non-copyable things (and possibly arrays) which is now fixed (see below)

> Regarding non-copyable things

I've opted for the simpler option for now. As you said, incremental changes FTW :) 
I think the main thing that needs addressing with the current solution is the duplication of `build_pod_safe_type_set`. I wasn't quite sure where to move it to. Do you have any suggestions?

> Getting references/pointers to fields - my first instinct is not to do it as part of this PR but do it later

That sounds reasonable!

adetaylor commented on engine/src/conversion/analysis/struct_accessors.rs:
```diff
@@ -0,0 +1,195 @@
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::collections::HashSet;
```
Please avoid using `HashSet` or `HashMap` and instead use some collection which has predictable iteration order. (I tend to use `use indexmap::set::IndexSet as HashSet`) so that we have deterministic behavior.

adetaylor commented:
> the duplication of build_pod_safe_type_set

Good question. How's about a new `api_utils.rs` alongside `api.rs`?

adetaylor commented:
This is looking pretty good now!

drozdziak1 commented:
Is there a reason not to merge this?

adetaylor commented:
That, plus it's still marked draft, it has a merge conflict (probably very easy to resolve) and there are still comments above that are not yet marked resolved. But I do want to merge this ASAP!

darichey commented:
Hi, sorry for the delay here. Unfortunately I don't have the time to continue with this PR. If anyone would like to pick it up and get it over the finish line, please feel free!

silvanshade commented:
I was able to update this PR branch with the suggested changes in a local fork of mine a little while ago and would be willing to submit a follow up PR with those fixes sometime soon.

Enduriel commented:
Sorry for necroing so badly, but this would be an incredible addition that seems to have stopped just before it was complete, and it would be incredibly useful for myself and certainly others. I was wondering if you were still able to and interested in getting it over the finish line @silvanshade. Would be awesome if so!

silvanshade commented:
@Enduriel I think this would be really cool to have but I probably won't return to work on it myself.

My efforts regarding Rust and C++ interop are currently focused on building Rust bindings to the LLVM C++ APIs (e.g., LibTooling and all the rest of it) in a different set of crates as an alternative `autocxx`.

Ultimately I think this is a more viable long term approach than trying to continue struggling with the limitations of bindgen and libclang and should make stuff like this easier eventually.

For what it's worth, the changes I had made to my local branch to get this to work were pretty minimal.

Enduriel commented:
Alright, unfortunately I don't have the experience here to make pretty much any useful changes, I actually ran into a problem I'll probably be writing another issue about when trying to use autocxx for my very specific use case.

FWIW, I wish you the best of luck and I hope this stuff does become easier because it's not been a great time for me personally 😅.
